src/python/gudhi/wasserstein/quantization.py

```diff
@@ -0,0 +1,189 @@
+# This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+# See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+# Author(s):       Theo Lacombe
+#
+# Copyright (C) 2021 Université Gustave Eiffel
+#
+# Modification(s):
+#   - YYYY/MM Author: Description of the modification
+
+
+import numpy as np
+from gudhi.wasserstein import _utils
+import warnings
+
+
+def _get_cells(X, c, withdiag, internal_p):
+    """
+    Given a point cloud `X` and a codebook (set of centroids) `c`,
+    assign to each point in X its nearest centroid in c.
+    If withdiag is True, points which are nearest to the diagonal than any centroid are assigned to an
+    additional cell.
+
+    Parameters
+    ----------
+    :param X: Current set of points, of size (n x d) (n : number of points, and d=2 for persistence diagrams).
+    :type X: `numpy.array`
+    :param c: Current set of centroids (codebook), of size (k x d). (k centroids in dimension d, d=2 for PDs).
+                It is assumed that k >= 1.
+    :param withdiag: Indicate if we consider the diagonal into account (True when working with PDs).
+    :type withdiag: bool
+    :param internal_p: The ground metric used to compute distances between points (norm_p).
+
+    :returns: list `cells` of affectation, such that cells[j] = points in `X` matched to `c[j]`,
+                with convention that `cells[k]` represents the diagonal whenever `withdiags=True`.
+
+        .. note: if the input X is empty, returns the empty list. Should not happen as _get_batches remove
+                    batches of empty points.
+
+    """
+    k = c.shape[0]
+
+    if X.shape[0] == 0:
+        warnings.warn("Input point cloud is empty, Cells affectation is a list of empty lists.")
+        return [[]] * (k+withdiag)  # k or (k+1) depending on additional centroid for the diagonal.
+
+    if X.shape[1] != 2:
+        raise ValueError("Input batch must be of shape (n x 2), not (%s,%s)" %(X.shape))
+
+    M = _utils._build_dist_matrix(X, c, order=2, internal_p=internal_p)  # Note: Order is useless here
+                                                                  # we just need nearest neighbors.
+
+    if withdiag:
+        a = np.argmin(M[:-1, :], axis=1)
+    else:
+        a = np.argmin(M[:-1, :-1], axis=1)
```

Comment:
It feels a bit strange to call `_build_dist_matrix`, whose main difference with `cdist` is that it adds the diagonal, just to drop the diagonal immediately... But I don't think it really matters.